src/Symfony/Component/PropertyAccess/PropertyAccessor.php

```diff
@@ -666,30 +676,69 @@ private function getWriteAccessInfo(string $class, string $property, $value): ar
-                );
             } else {
-                $access[self::ACCESS_TYPE] = self::ACCESS_TYPE_NOT_FOUND;
-                $access[self::ACCESS_NAME] = sprintf(
-                    'Neither the property "%s" nor one of the methods %s"%s()", "%s()", '.
-                    '"__set()" or "__call()" exist and have public access in class "%s".',
-                    $property,
-                    implode('', array_map(function ($singular) {
-                        return '"add'.$singular.'()"/"remove'.$singular.'()", ';
-                    }, $singulars)),
-                    $setter,
-                    $getsetter,
-                    $reflClass->name
-                );
+                foreach ($this->findAdderAndRemover($reflClass, $singulars) as $methods) {
+                    if (3 === \count($methods)) {
+                        $errors[] = sprintf(
+                            'The property "%s" in class "%s" can be defined with the methods "%s()" but '.
+                            'the new value must be an array or an instance of \Traversable, '.
+                            '"%s" given.',
+                            $property,
+                            $reflClass->name,
+                            implode('()", "', [$methods[self::ACCESS_ADDER], $methods[self::ACCESS_REMOVER]]),
+                            \is_object($value) ? \get_class($value) : \gettype($value)
+                        );
+                    }
+                }
+
+                if (!isset($access[self::ACCESS_NAME])) {
+                    $access[self::ACCESS_TYPE] = self::ACCESS_TYPE_NOT_FOUND;
+
+                    $triedMethods = [
+                        $setter => 1,
+                        $getsetter => 1,
+                        '__set' => 2,
+                        '__call' => 2,
+                    ];
+
+                    foreach ($singulars as $singular) {
+                        $triedMethods['add'.$singular] = 1;
+                        $triedMethods['remove'.$singular] = 1;
+                    }
+
+                    foreach ($triedMethods as $methodName => $parameters) {
+                        if (!$reflClass->hasMethod($methodName)) {
+                            continue;
+                        }
+
+                        $method = $reflClass->getMethod($methodName);
+
+                        if (!$method->isPublic()) {
+                            $errors[] = sprintf('The method "%s" in class "%s" was found but does not have public access', $methodName, $reflClass->name);
+                            continue;
+                        }
+
+                        if ($method->getNumberOfRequiredParameters() > $parameters || $method->getNumberOfParameters() < $parameters) {
+                            $errors[] = sprintf('The method "%s" in class "%s" requires %d arguments, but should accept only %d', $methodName, $reflClass->name, $method->getNumberOfRequiredParameters(), $parameters);
```

Comment:
IIUC (but I might be completely wrong):

`requires %d arguments` -> that's the number of passed values right? What about `was passed %d values, but you. must pass exactly %s arguments`?

Reply:
This is when a method is defined with the incorrect number of parameters. E.G. the method is defined as `function addFoo()`, but it needs to be defined as `function addFoo($var)`. So this message tells the user that the method is currently defined to require x arguments, but it should be defined to only require y arguments.

Reply:
@pierredup Can you change the current message to make it more explicit?